### PR TITLE
Avoid requiring identifiable for FieldValue

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR
I will reduce the size back to what it was afterwards in a separate PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/memoryindex/url_field_inverter.cpp [hunk 1]
```diff
@@ -309,30 +307,29 @@ UrlFieldInverter::invertUrlField(const FieldValue &val)
...
-    case CollectionType::WEIGHTEDSET:
-        if (cInfo.id() == WeightedSetFieldValue::classId) {
-            const auto &wset = static_cast<const WeightedSetFieldValue &>(val);
-            if (isUriType(wset.getNestedType())) {
-                processWeightedSetUrlField(wset);
-            } else {
-                throw std::runtime_error(make_string("Expected wset of URI struct, got '%s'", wset.getNestedType().getName().c_str()));
-            }
+    case CollectionType::WEIGHTEDSET: {
+        assert(val.isA(FieldValue::Type::WSET));
```
Comment (vekterli) [inline searchlib/src/vespa/searchlib/memoryindex/url_field_inverter.cpp:311]: Should these checks use `type() == ` instead, or conversely, should the `type() ==` checks earlier in this file use `isA()`?

File: vsm/src/vespa/vsm/vsm/slimefieldwriter.cpp [hunk 2]
```diff
@@ -48,21 +49,21 @@ namespace vsm {
 void
 SlimeFieldWriter::traverseRecursive(const document::FieldValue & fv, Inserter &inserter)
 {
-    const auto & clazz = fv.getClass();
     LOG(debug, "traverseRecursive: class(%s), fieldValue(%s), currentPath(%s)",
-        clazz.name(), fv.toString().c_str(), toString(_currPath).c_str());
+      vespalib::getClassName(fv).c_str(), fv.toString().c_str(), toString(_currPath).c_str());
```
Comment (vekterli) [inline vsm/src/vespa/vsm/vsm/slimefieldwriter.cpp:53]: Nit: indent is slightly off

File: document/src/vespa/document/fieldvalue/fieldvalue.h [hunk 3]
```diff
@@ -13,38 +13,31 @@
...
 {
-protected:
-    FieldValue(const FieldValue&) = default;
-    FieldValue& operator=(const FieldValue&) = default;
-    FieldValue(FieldValue &&) = default;
-    FieldValue& operator=(FieldValue &&) = default;
-    static std::unique_ptr<vespalib::IArrayBase> createArray(const DataType & baseType);
-
 public:
+    enum class Type : uint8_t {NONE,BOOL, BYTE, SHORT, INT, LONG, FLOAT, DOUBLE, STRING, RAW, PREDICATE, TENSOR, ANNOTATION_REFERENCE, REFERENCE, ARRAY, WSET, MAP, STRUCT, DOCUMENT};
```
Comment (vekterli) [inline document/src/vespa/document/fieldvalue/fieldvalue.h:33]: Nit: missing spacing after `NONE,`. Also, consider splitting over multiple lines

File: document/src/vespa/document/update/addfieldpathupdate.cpp [hunk 4]
```diff
@@ -53,13 +54,13 @@ class AddIteratorHandler : public fieldvalue::IteratorHandler {
...
 AddIteratorHandler::doModify(FieldValue &fv) {
-    if (fv.inherits(CollectionFieldValue::classId)) {
+    if (fv.isCollection()) {
         CollectionFieldValue &cf = static_cast<CollectionFieldValue &>(fv);
         for (std::size_t i = 0; i < _values.size(); ++i) {
             cf.add(_values[i]);
         }
     } else {
-        vespalib::string err = make_string("Unable to add a value to a \"%s\" field value.", fv.getClass().name());
+        vespalib::string err = make_string("Unable to add a value to a \"%s\" field value.", vespalib::getClassName(fv).c_str());
```
Comment (vekterli) [inline document/src/vespa/document/update/addfieldpathupdate.cpp:63]: It feels slightly naughty to me to invoke RTTI and ABI-specific demangling to get to the class names of a small set of concrete types. Should we instead have a utility function on `FieldValue` to get the class/type name, either virtual or just a big ol' `switch` block based on `FieldValue::_type`?

**Review by vekterli — APPROVED** [review]
:+1:

File: document/src/vespa/document/fieldvalue/fieldvalue.h [hunk 5]
```diff
@@ -174,17 +167,39 @@ class FieldValue : public vespalib::Identifiable
...
+    Type type() const { return _type; }
+    bool isA(Type type) const { return type == _type; }
+    bool isCollection() const { return (_type == Type::WSET) || (_type == Type::ARRAY); }
+    bool isStructured() const { return (_type == Type::DOCUMENT) || (_type == Type::STRUCT); }
+    bool isLiteral() const { return (_type == Type::STRING) || (_type == Type::RAW); }
+    bool isNumeric() const {
+        return (_type == Type::BYTE) || (_type == Type::SHORT) ||
+               (_type == Type::INT) || (_type == Type::LONG) || (_type == Type::FLOAT) || (_type == Type::DOUBLE);
+    }
+    bool isFixedSizeSingleValue() const {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/fieldvalue.h:180]: Fixed

File: document/src/vespa/document/fieldvalue/fieldvalue.h [hunk 6]
```diff
@@ -174,17 +167,39 @@ class FieldValue : public vespalib::Identifiable
 
     // Utility functions to set commonly used value types.
     virtual FieldValue& operator=(vespalib::stringref);
+
+    Type type() const { return _type; }
+    bool isA(Type type) const { return type == _type; }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/fieldvalue.h:172]: No, that code uses isStructured instead()

File: document/src/vespa/document/fieldvalue/literalfieldvalue.h [hunk 7]
```diff
@@ -24,15 +24,14 @@ class LiteralFieldValueB : public FieldValue {
 public:
     typedef vespalib::string string;
     typedef vespalib::stringref stringref;
-    DECLARE_IDENTIFIABLE_ABSTRACT(LiteralFieldValueB);
     typedef std::unique_ptr<LiteralFieldValueB> UP;
     typedef string value_type;
 
-    LiteralFieldValueB();
+    LiteralFieldValueB(Type type);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/literalfieldvalue.h:30]: Done

File: document/src/vespa/document/fieldvalue/literalfieldvalue.h [hunk 8]
```diff
@@ -83,15 +82,15 @@ class LiteralFieldValueB : public FieldValue {
...
+template<typename SubClass, int dataType, bool addZeroTerm>
 class LiteralFieldValue : public LiteralFieldValueB {
 private:
     bool getAddZeroTerm() const  override{ return addZeroTerm; }
 public:
     typedef std::unique_ptr<SubClass> UP;
 
-    LiteralFieldValue() : LiteralFieldValueB() { }
-    LiteralFieldValue(const stringref& value) : LiteralFieldValueB(value) { }
+    LiteralFieldValue(Type type) : LiteralFieldValueB(type) { }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/literalfieldvalue.h:92]: Done

File: document/src/vespa/document/update/mapvalueupdate.cpp [hunk 9]
```diff
@@ -44,7 +44,7 @@ MapValueUpdate::checkCompatibility(const Field& field) const
 {
     // Check compatibility of nested types.
     if (field.getDataType().isArray()) {
-	    if (_key->getClass().id() != IntFieldValue::classId) {
+	    if (_key->type() != FieldValue::Type::INT) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/mapvalueupdate.cpp:47]: Fixed

File: document/src/vespa/document/update/tensor_modify_update.cpp [hunk 10]
```diff
@@ -116,7 +117,7 @@ TensorModifyUpdate::operator=(const TensorModifyUpdate &rhs)
         _operation = rhs._operation;
         _tensor.reset();
         _tensorType = std::make_unique<TensorDataType>(*rhs._tensorType);
-        _tensor.reset(Identifiable::cast<TensorFieldValue *>(_tensorType->createFieldValue().release()));
+        _tensor.reset(dynamic_cast<TensorFieldValue *>(_tensorType->createFieldValue().release()));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/tensor_modify_update.cpp:120]: Identifiable::cast behaves as dynamic_cast.
The places where it has been replaced with a static cast it is checked type with isA or similar.

File: document/src/vespa/document/update/valueupdate.h [hunk 11]
```diff
@@ -33,12 +33,13 @@ class DataType;
 
 class ValueUpdate : public vespalib::Identifiable,
                     public Printable,
-                    public XmlSerializable
+public vespalib::xml::XmlSerializable
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/update/valueupdate.h:36]: Fixed

File: searchlib/src/vespa/searchlib/memoryindex/url_field_inverter.cpp (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/memoryindex/url_field_inverter.cpp:311]: Fixed

File: vsm/src/vespa/vsm/vsm/slimefieldwriter.cpp (hunk 2 above)
Comment (baldersheim) [reply vsm/src/vespa/vsm/vsm/slimefieldwriter.cpp:53]: Fixed

File: document/src/vespa/document/fieldvalue/fieldvalue.h (hunk 3 above)
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/fieldvalue.h:33]: Fixed

File: document/src/vespa/document/update/addfieldpathupdate.cpp (hunk 4 above)
Comment (baldersheim) [reply document/src/vespa/document/update/addfieldpathupdate.cpp]: First option chosen with a switch block.

**Comment by baldersheim** [comment]
@vekterli PTAL

**Review by vekterli — APPROVED** [review]
:+1:

File: document/src/vespa/document/fieldvalue/fieldvalue.cpp [hunk 12]
```diff
@@ -26,8 +26,50 @@ namespace document {
...
+            return "WSetFieldValue";
+        case Type::MAP:
+            return "MapFieldValue";
+        case Type::STRUCT:
+            return "StructFieldValue";
+        case Type::DOCUMENT:
+            return "DocumentFieldValue";
+        case Type::NONE:
+        default:
+            return "NONE - Impossible";
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/vespa/document/fieldvalue/fieldvalue.cpp]: Done

**Pull request merged**